### PR TITLE
bpo-46267: Correctly use compresslevel in gzip.compress 

### PR DESCRIPTION
**Pull request by rhpvorderman**
This is a small one-liner bug fix. I am the one responsible for introducing the bug in the first place, so I am happy I found it before 3.11 shipped.

https://bugs.python.org/issue46267

#90425

### CODE REVIEW

**Comment by rhpvorderman** [comment]
No news item is required. This new method for gzip.compress was introduced in 3.11.

File: Lib/gzip.py
Comment (akulakov) [inline Lib/gzip.py]: According to the docs, `level` arg is keyword-only.

[Edit: I was wrong about this, please ignore]

Comment (akulakov) [reply Lib/gzip.py:590]: Hmm, the docs seem to be wrong though. It works as positional. However on line 586 it's used as kw arg, so it's a bit better to do that for consistency.

Comment (rhpvorderman) [reply Lib/gzip.py]: Since you thoroughly reviewed whether this one word is correctly added to the code, may I kindly request that you push the merge button? Thank you for your review!

Comment (akulakov) [reply Lib/gzip.py]: Can you look into adding a unit test for this?

I can't merge the PR (not a core dev), but I can label it for core dev review, if you can add a unit test.

Comment (rhpvorderman) [reply Lib/gzip.py]: No I can't. A raw DEFLATE stream does not contain any information on its compression level. And that level 9 is smaller than level 1 is not a given depending on the test data. So there is not a way to reliable ascertain that the correct compression level is used.

What I can test is whether the XFL is set correctly, but that is already in the test suite. That is a field in the gzip header though, so it is not affected by this change.

Comment (akulakov) [reply Lib/gzip.py]: I believe that what you can do here is strip header (10 chars for data i tested), and compare the remaining data between two runs with differing compression levels. If that returned data is different, the test should pass.

Comment (rhpvorderman) [reply Lib/gzip.py]: You got me thinking. I made a false statement here:

> No I can't. A raw DEFLATE stream does not contain any information on its compression level. 

Because raw deflate does contain a `BTYPE` that notes what kind of stream it is. And there is a special btype for an uncompressed stream (compression level 0), so it could indeed be tested. I will add it right away.

Comment (rhpvorderman) [reply Lib/gzip.py]: A regression test was added. This takes advantage of the fact that a stream compressed with compresslevel 0 , should still contain the original data.

File: Lib/test/test_gzip.py
Comment (akulakov) [inline Lib/test/test_gzip.py]: It's better to use `assertIn` and `assertNotIn`.

**Comment by akulakov** [comment]
This does require a news entry if the minor version is different vs. where bug was introduced. This is true for all bugfixes.

**Comment by rhpvorderman** [comment]
> This does require a news entry if the minor version is different vs. where bug was introduced. This is true for all bugfixes.

This bug was introduced in 3.11. I am hoping this will still get into main before python 3.11 releases. So no news item is required.

File: Lib/test/test_gzip.py
Comment (rhpvorderman) [reply Lib/test/test_gzip.py]: Thanks. I have changed it.

**Comment by akulakov** [comment]
> > This does require a news entry if the minor version is different vs. where bug was introduced. This is true for all bugfixes.
> 
> This bug was introduced in 3.11. I am hoping this will still get into main before python 3.11 releases. So no news item is required.

3.11 is a major release, what was the minor release number?

**Comment by rhpvorderman** [comment]
> 3.11 is a major release, what was the minor release number?

Ah, I see I was using semantic versioning terminology, but obviously that does not apply here.
The bug was introduced in Python 3.11.0 alpha 1.

How shall I describe the news entry? It is a bit silly that the 3.11 final released version will have two entries describing the introduction of a performance improvement and a bugfix in it. So how would I word this?

**Comment by rhpvorderman** [comment]
A news entry was added.

File: Misc/NEWS.d/next/Library/2022-01-18-08-03-19.bpo-46267.vbmuum.rst
Comment (akulakov) [inline Misc/NEWS.d/next/Library/2022-01-18-08-03-19.bpo-46267.vbmuum.rst]: spell 'through'

**Comment by akulakov** [comment]
Thanks for adding it, I think we are almost there. 
It's useful to have two news entries for the same reason you would have them if feature was added in 3.10 and bugfix in 3.11 -- users can have a version installed with the feature but before bugfix and we should tell them there's a bug and which exact version to upgrade to if the bug affects them.

File: Misc/NEWS.d/next/Library/2022-01-18-08-03-19.bpo-46267.vbmuum.rst
Comment (rhpvorderman) [reply Misc/NEWS.d/next/Library/2022-01-18-08-03-19.bpo-46267.vbmuum.rst]: Thanks. Fixed.

**Comment by akulakov** [comment]
LGTM!

**Comment by rhpvorderman** [comment]
I squashed the commits. 4 commits for 12 lines was a bit too much. The code content is still the same.

**Comment by JelleZijlstra** [comment]
#31215 already fixed this, but didn't add tests. I'll merge this PR too to make sure we have tests.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by rhpvorderman** [comment]
> https://github.com/python/cpython/pull/31215 already fixed this, [...]

No, this PR already fixed it. Five months ago, four months before that other PR. It has had the bugfix label for 5 months. It is a bit unfortunate that not merging it sooner eventually lead to duplicated effort. But I understand keeping up is hard with the limited amount of people available.

Anyway, I am glad it is in now. I also happened to be the one who introduced the bug (when I made the compress function have quite a lot less overhead). Imagine the shame of having broken something in CPython. Thanks!

**Comment by JelleZijlstra** [comment]
> > #31215 already fixed this, [...]
> 
> No, this PR already fixed it. Five months ago, four months before that other PR. It has had the bugfix label for 5 months. It is a bit unfortunate that not merging it sooner eventually lead to duplicated effort. But I understand keeping up is hard with the limited amount of people available.

I'm sorry for the imprecise wording there. There is a lot of open PRs and it's difficult to keep up.

**Comment by JelleZijlstra** [comment]
Thanks for your contribution, and sorry again this PR had to wait for so long!

**Comment by rhpvorderman** [comment]
Sorry for my sour initial response. As you can see I had to jump trough quite a lot of hoops to satisfy alukov for this simple one liner. Not that I mind, the PR became much better for it. But for this PR to close after 5 months due to someone else getting a oneliner fix in that did not get the same treatment... Well, it is quite an anti-climax ;-).

Thanks again for merging and I hope you have a nice day!

**Pull request merged**
